README.md
# Dollhouse

![](doc_image/35.png)<br />

![](doc_image/36.png)<br />

## Description

First large project I've attempted using CadQuery. <br />
Goal was to make a modular dollhouse for my kids.

![](doc_image/33.png)<br />


It's a tudor style house with inspiration drawn from google image searches.<br />
The central idea is that it would be printed in roughly nine sections. With the floors and roof divided into three sections each.

I opted for the main rooms to be 175mm x 175mm x 175mm or roughly 7" x 7" x 7".<br />
Which is about the limits of my print bed.

---
## Code Overview

### Dependencies
* [CadQuery](https://cadquery.readthedocs.io/en/latest/) - Python Cad Libary
* [CQ-Editor](https://github.com/CadQuery/CQ-editor) - GUI to see model changes
* [cadqueryhelper](https://github.com/medicationforall/cadqueryhelper) - Shape primitives libraries, used for object repetition. (I wrote this)
* [cqterrain](https://github.com/medicationforall/cqterrain) - Primary library used to build the house (I authored this as well)

I opted for python code to make the cad model because I like making 3d models with code.<br />I've used [OpenSCAD](https://openscad.org/) in the past, but it has limitations for large projects.

### Pros
* Python has a large ecosystem
* Parts of the model can be broken up into smaller modules and tested in solation.
* Version control - I can review changes to the model before committing.
* [Github](https://github.com/) - code is kept in the cloud in a private (or public) repo.
* Re-using code from one project to the next is pretty straight forward.

### Cons
* Tedious, the project itself was 500+ lines of code just for the dollhouse itself.
* The re-usable api was expanded out as the model was being created, which slowed down development.
* If there are bugs in your logic tracking them down can be arduous, the generated models do not lend themselves well to unit tests.

---

### In The Beginning
The initial outline of the project.
![](doc_image/04.png)<br />
The front is too boring

![](doc_image/03.png)<br />
Shows the breakdown of the nine sections.

---

### Adding Exterior Details
![](doc_image/06.png)<br />
I ended up simplifying from the initial sketch.

![](doc_image/05.png)<br />
I used [Microsoft 3d Viewer](https://all3dp.com/2/microsoft-3d-viewer-guide/) to generate the lighting on the model

![](doc_image/07.png)<br />
Interior is still plain

* Arches
* Stones
* Casement Windows
* Lattice Windows
* Roof

---

## Arches
The arch cutouts are straight forward.<br />
``` python
def make_arch_door(wall, length, width, height, floor_height):
    # find the bottom of the wall to align to.
    bottom = wall.faces("-Z").val()

    #create the initial shape
    cutout = (cq.Workplane(bottom.Center())
              .box(length, width, height)
              .translate((0,0,(height/2)+floor_height))
              )

    # round off the top          
    cutout = cutout.faces("Z").edges("Y").fillet((length/2)-.5)

    #remove the arch from the wall
    w = wall.cut(cutout)
    return w
```
The arch is aligned to the bottom of the object it's being cut out of.
![](doc_image/34.png)<br />


### Stones
I opted, to write a pseudo-random stone pattern generator
![](doc_image/stone_01.png)<br />

Code for a stone section.

``` python
def add_stones(wall, length, height, wall_width, rotate=0, seed="test4"):
    # static boxes to act as stone
    tile = cq.Workplane("XY").box(10,10,2)
    tile2 = cq.Workplane("XY").box(8,8,2)
    tile3 = cq.Workplane("XY").box(6,12,2)

    # create an array of the stones and chamfer / fillet to add interest to the shapes.
    stone_list = [tile.chamfer(0.8), tile2.fillet(.5), tile3.chamfer(0.5)]

    # This is the pattern generator
    stones = stone.make_stones(stone_list, [12,12,2], columns = 14, rows = 3, seed=seed).rotate((0,1,0),(0,0,0), 90).rotate((0,0,1),(0,0,0), 90)

    # Align the pattern and surround with a frame
    stones = stones.translate((0,-2,-1*(height/2)+(24))).rotate((0,0,1),(0,0,0), rotate)
    frame = window.frame(length, 2, 48).translate((0,-1*((1)+(wall_width/2)),-1*(height/2)+(24))).rotate((0,0,1),(0,0,0), rotate)

    # Add the detailing to the room wall
    return wall.add(stones).add(frame)
```

* The generated output of the stone pattern is defined by the seed.
  * different seed means different stone placement.

Let's looks at cqterrain [stone.make_stones](https://github.com/medicationforall/cqterrain/blob/main/src/cqterrain/stone.py) code.

``` python
import cadquery as cq
import random
import math

def make_stones(parts, dim=[5,5,2], rows=2, columns=5, seed="test4"):
    grid = cq.Assembly()
    random.seed(seed)

    # loop the rows
    for row_i in range(rows):
        row_offset = (dim[0] * row_i)
        # loop the columns per row
        for col_i in range(columns):
            col_offset = (dim[1] * col_i)

            col_push_x = 0
            col_push_y = 0
            if col_i % 2 == 1:
                col_push_x = 0
                col_push_y = 0

            z_push=0
            # move the part in a random direction along the x and y axis.
            x_rand = random.randrange(-1*(math.floor(dim[0]/2)),(math.floor(dim[0]/2)))
            y_rand = random.randrange(-1*(math.floor(dim[1]/2)),(math.floor(dim[1]/2)))

            # choose a random part from the parts list
            part_index = random.randrange(0,len(parts))

            # add the part to the assembly
            grid.add(parts[part_index], loc=cq.Location(cq.Vector(row_offset + col_push_x + x_rand, col_offset + col_push_y + y_rand, z_push)))

    length = dim[1] * columns
    width = dim[0] * rows

    # dump the assembly out as a single compound
    comp = grid.toCompound()
    work = cq.Workplane("XZ").center(0, 0).workplane()
    work.add(comp)

    # zero out the grid
    work = work.translate(((dim[0]/2),(dim[1]/2)))
    work = work.translate((-1*(width/2),-1*(length/2)))
    return work
```

* That's all of of it, the double for loop isn't ideal but I kept it for readability.
* I basically yoinked my code for laying a tile onto a grid, and modified it to support pseudo-random selection and placement.
* It's not perfect but it's good enough to give the impression of a stone pattern.
* The code isn't resource intensive since it's made up of very simple primitives.

![](doc_image/39.png)<br />

## Casement Windows
![](doc_image/window_casement_01.png)<br />

Code for making the windows.
``` python
def casement_windows(wall, length, width, height, count, padding):
    # Create the cut out the holes where the windows will be placed.
    window_cutout = cq.Workplane().box(length, width, height)
    window_cut_series = series(window_cutout, count, length_offset = padding)

    # create the window frame, and grill.
    i_window = window.frame(length, width+3, height)
    grill = window.grill(length=length, width=4, height=height, rows=4, columns=2, grill_width=2, grill_height=3)
    i_window.add(grill)

    # Create the window set
    window_series = series(i_window, count, length_offset = padding)

    # Remove the cutout and add the windows
    w = wall.cut(window_cut_series)
    w = w.add(window_series)

    return w
```

Making the grill.

``` python
def grill(length=20, width=4, height=40, columns=4, rows=2, grill_width=1, grill_height=1):
    # Make a flat plane
    pane = cq.Workplane("XY").box(length, grill_height, height)
    t_width = length / columns
    t_height = height / rows

    # Make the window cutout
    tile = cq.Workplane("XY").box(t_width, grill_height, t_height).rotate((1,0,0),(0,0,0),90)

    # Repeat the cutout
    tiles = grid.make_grid(tile, [t_width+grill_width, t_height+grill_width], rows=columns, columns=rows).rotate((1,0,0),(0,0,0),-90)

    # Remove the window cutouts leaving the frame
    combine = pane.cut(tiles)
    return combine
```

![](doc_image/window_casement_02.png)<br />
The tudor framing on the outside of the house are just these casement window grills.

![](doc_image/38.png)<br />


### Lattice Windows
![](doc_image/window_lattice_01.png)<br />

The only difference between the casement and the lattice is the grill pattern.

``` python
def lattice(length=20, width=4, height=40,  tile_size=4, lattice_width=1, lattice_height=1, lattice_angle=45):
    # Determine longest distance between points
    hyp = math.hypot(length, height)
    columns= math.floor(hyp / (tile_size+lattice_width))
    rows= math.floor(hyp / (tile_size+lattice_width))

    # Make a flat plane
    pane = cq.Workplane("XY").box(length, lattice_height, height)

    #make the cutout tile
    tile = cq.Workplane("XY").box(tile_size, lattice_height, tile_size).rotate((1,0,0),(0,0,0),90)
    tiles = grid.make_grid(tile, [tile_size+lattice_width, tile_size+lattice_width], rows=columns, columns=rows).rotate((1,0,0),(0,0,0),-90).rotate((0,1,0),(0,0,0),lattice_angle)
    combine = pane.cut(tiles)
    return combine
```
![](doc_image/40.png)<br />



### Roof
![](doc_image/roof_08.png)<br />
The roof tiles were a struggle but it was a good opportunity re-learn some trigonometry.

Code to make a roof.
``` python
def make_roof(roof_width=185, x_offset=0):
    # Make the wedge shape
    gable_roof_raw = roof.dollhouse_gable(length=roof_width, width=185, height=100)

    # Shell the roof to cut out the inside
    gable_roof = roof.shell(gable_roof_raw,face="Y", width=-4)

    # Determine the arccosine angle of the roof
    angle = roof.angle(185, 100)
    face_x = gable_roof_raw.faces("<X")

    # Feature to enable/disable rendering roof tiles
    if render_roof_tiles:
        # Individual roof tile
        tile = cq.Workplane("XY").box(15,12,2).rotate((0,1,0),(0,0,0),8)
        # Grid of tiles
        tiles = roof.tiles(tile, face_x, 185, 100, 15, 12, angle, rows=28, odd_col_push=[3,0], intersect=False).rotate((0,0,1),(0,0,0),90).translate((3,45,0))
        tiles = tiles.translate((x_offset,0,0))

        # Cut away box to remove excess tiles
        inter_tiles = cq.Workplane("XY").box(roof_width,185, 100)
        inter_tiles = tiles.intersect(inter_tiles)
        return gable_roof.add(inter_tiles)
    else:
        # Quick roof no tiles
        return gable_roof
```
Making the tiles is resource intensive, so a feature flag was added for quick rendering.

Making the wedge
``` python
def dollhouse_gable(length= 40, width=40, height=40):

Changes.md
# dollhouse changelog

## Main wip
* Updated License

## 1.0.0
* Initial release


dollhouse.py
# Copyright 2022 James Adams
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import cadquery as cq
from cqterrain import Building, window, roof, stone, stairs, Ladder, tile
from cadqueryhelper import series, grid
import math

cq_editor_show = True
export_to_file = False
render_floor = True
render_roof_tiles = True

def test_operation(f):
    box = cq.Workplane("XY").box(31,40,30).translate((-42.5,44.5,0))
    f = f.cut(box)
    return f

def add_tudor_frame(wall, length, height, wall_width, frame_width=5, frame_height=5, rows=2, columns = 5,  w_length=0, w_height=0, rotate=0):
    #show_object(wall)
    t_length = length/ columns - frame_width
    t_width = frame_height
    t_height =  (height / rows) - frame_width

    frame = window.frame(length, frame_height, height).translate((0,-1*((frame_height/2)+(wall_width/2)),0))
    grill = window.grill(length, wall_width, height, columns, rows, frame_width, frame_height ).translate((0,-1*((frame_height/2)+(wall_width/2)),0))
    if w_length and w_height:
        win_cut = cq.Workplane("XY").box(w_length, frame_height, w_height).translate((0,-1*((frame_height/2)+(wall_width/2)),0))

        grill_cut = grill.cut(win_cut)
        #show_object(grill_cut)

        if rotate:
            grill_cut = grill_cut.rotate((0,0,1),(0,0,0),rotate)
            frame = frame.rotate((0,0,1),(0,0,0),rotate)
        return wall.add(frame).add(grill_cut)
    #show_object(grill2)
    if rotate:
        grill = grill.rotate((0,0,1),(0,0,0),rotate)
        frame = frame.rotate((0,0,1),(0,0,0),rotate)
    return wall.add(frame).add(grill)

def add_stones(wall, length, height, wall_width, rotate=0, seed="test4"):
    tile = cq.Workplane("XY").box(10,10,2)
    tile2 = cq.Workplane("XY").box(8,8,2)
    tile3 = cq.Workplane("XY").box(6,12,2)
    stone_list = [tile.chamfer(0.8), tile2.fillet(.5), tile3.chamfer(0.5)]
    stones = stone.make_stones(stone_list, [12,12,2], columns = 14, rows = 3, seed=seed).rotate((0,1,0),(0,0,0), 90).rotate((0,0,1),(0,0,0), 90)
    stones = stones.translate((0,-2,-1*(height/2)+(24))).rotate((0,0,1),(0,0,0), rotate)
    frame = window.frame(length, 2, 48).translate((0,-1*((1)+(wall_width/2)),-1*(height/2)+(24))).rotate((0,0,1),(0,0,0), rotate)

    #show_object(stones)
    #show_object(wall)
    #show_object(frame)
    return wall.add(stones).add(frame)



def make_roof(roof_width=185, x_offset=0):
    # Make the wedge shape
    gable_roof_raw = roof.dollhouse_gable(length=roof_width, width=185, height=100)

    # Shell the roof to cut out the inside
    gable_roof = roof.shell(gable_roof_raw,face="Y", width=-4)

    # Determine the arccosine angle of the roof
    angle = roof.angle(185, 100)
    face_x = gable_roof_raw.faces("<X")

    # Feature to enable/disable rendering roof tiles
    if render_roof_tiles:
        # Individual roof tile
        tile = cq.Workplane("XY").box(15,12,2).rotate((0,1,0),(0,0,0),8)
        # Grid of tiles
        tiles = roof.tiles(tile, face_x, 185, 100, 15, 12, angle, rows=28, odd_col_push=[3,0], intersect=False).rotate((0,0,1),(0,0,0),90).translate((3,45,0))
        tiles = tiles.translate((x_offset,0,0))

        # Cut away box to remove excess tiles
        inter_tiles = cq.Workplane("XY").box(roof_width,185, 100)
        inter_tiles = tiles.intersect(inter_tiles)
        return gable_roof.add(inter_tiles)
    else:
        # Quick roof no tiles
        return gable_roof


def make_dormer_roof(roof_part, width=185):
    gable_roof_raw = roof.dollhouse_gable(length=width, width=185, height=100).translate((0,0,-4.5))

    length=185
    height = 100
    inner_height = 60
    roof_half_one = roof.dollhouse_gable(length=140, width=40, height=30).translate((0,0,29)).rotate((0,0,1),(0,0,0),90).translate((-20,15,0))
    roof_half_two = roof.dollhouse_gable(length=140, width=40, height=30).translate((0,0,29)).rotate((0,0,1),(0,0,0),-90).translate((20,15,0))

    if render_roof_tiles:
        angle = roof.angle(40, 30)
        face_x = roof_half_one.faces(">X")
        tile = cq.Workplane("XY").box(15,12,2).rotate((0,1,0),(0,0,0),8)
        tiles = roof.tiles(tile, face_x, 140, 30, 15, 12, angle, rows=4, odd_col_push=[3,0], intersect=False).translate((-14.5,23,29))
        tiles2 = roof.tiles(tile, face_x, 140, 30, 15, 12, angle, rows=4, odd_col_push=[3,0], intersect=False).translate((-14.5,23,29)).rotate((0,0,1),(0,0,0),180).translate((0,46,0))


    inner = roof_part.faces("<Z").box(80,110,inner_height, combine=False).translate((0,0,inner_height/2+4))
    inner = inner.union(roof_half_one).union(roof_half_two)
    #show_object(inner)
    #show_object(roof_part)

    inner_shell = roof_part.faces("<Z").box(80,140,inner_height, combine=False).translate((0,15,inner_height/2+4))
    inner_shell = inner_shell.union(roof_half_one).union(roof_half_two)

    inner_shell = inner_shell.faces(">Y").shell(-4)

    if render_roof_tiles:
        tile_cut = cq.Workplane("XY").box(40,140,50).translate((20,15,25))
        tile_cut2 = cq.Workplane("XY").box(40,140,50).translate((-20,15,25))

        tiles = tiles.intersect(tile_cut)
        tiles2 = tiles2.intersect(tile_cut2)

        inner_shell = inner_shell.add(tiles).add(tiles2)
    inner_shell = inner_shell.cut(gable_roof_raw)

    combine = roof_part.cut(inner).add(inner_shell)

    window_slug = inner.faces("<Y").cylinder(8,20,combine=False).rotateAboutCenter((1,0,0),90).translate((0,2.5,10))
    window_inner = inner.faces("<Y").cylinder(8,17,combine=False).rotateAboutCenter((1,0,0),90).translate((0,2.5,10))
    win_frame = window_slug.cut(window_inner)
    grill = window.grill(40, 5, 40, 2, 2, 3, 3 ).translate((0,-52,5))
    combine = combine.cut(window_slug).add(win_frame).add(grill)

    #show_object(window_slug)
    #show_object(combine)
    return combine



def lattice_windows(wall, length, width, height, count, padding):
    # Create the cut out the holes where the windows will be placed.
    window_cutout = cq.Workplane().box(length, width, height)
    window_cut_series = series(window_cutout, count, length_offset = padding)

    # create the window frame, and lattice.
    i_window = window.frame(length, width+3, height)
    lattice = window.lattice(length=length, width=4, height=height, tile_size=6, lattice_height=2)
    i_window.add(lattice)

    # Create the window set
    window_series = series(i_window, count, length_offset = padding)

    # remove the cutout and add the windows
    w = wall.cut(window_cut_series)
    w = w.add(window_series)

    return w

def casement_windows(wall, length, width, height, count, padding):
    # Create the cut out the holes where the windows will be placed.
    window_cutout = cq.Workplane().box(length, width, height)
    window_cut_series = series(window_cutout, count, length_offset = padding)

    # create the window frame, and grill.
    i_window = window.frame(length, width+3, height)
    grill = window.grill(length=length, width=4, height=height, rows=4, columns=2, grill_width=2, grill_height=3)
    i_window.add(grill)

    # Create the window set
    window_series = series(i_window, count, length_offset = padding)

    # remove the cutout and add the windows
    w = wall.cut(window_cut_series)
    w = w.add(window_series)

    return w

def casement_windows_2(wall, length, width, height, count, padding):
    window_cutout = cq.Workplane().box(length, width, height)
    window_cut_series = series(window_cutout, count, length_offset = padding)

    i_window = window.frame(length, width+3, height)
    grill = window.grill(length=length, width=4, height=height, rows=4, columns=4, grill_width=2, grill_height=3)
    i_window.add(grill)
    window_series = series(i_window, count, length_offset = padding)

    w = wall.cut(window_cut_series)
    w = w.add(window_series)

    return w


def make_arch_door(wall, length, width, height, floor_height):
    # find the bottom of the wall to align to.
    bottom = wall.faces("-Z").val()

    #create the initial shape
    cutout = (cq.Workplane(bottom.Center())
              .box(length, width, height)
              .translate((0,0,(height/2)+floor_height))
              )

    # round off the top
    cutout = cutout.faces("Z").edges("Y").fillet((length/2)-.5)
    #remove the arch from the wall

    w = wall.cut(cutout)
    return w

def make_arch_door_fancy(wall, length, width, height, floor_height):
    bottom = wall.faces("-Z").val()
    cutout = (cq.Workplane(bottom.Center())
              .box(length, width+2, height)
              .translate((0,2,(height/2)+floor_height))
              )
    cutout = cutout.faces("Z").edges("Y").fillet((length/2)-.5)

    door_frame = (cq.Workplane(bottom.Center())
              .box(length+10, width+2, height+10)
              .translate((0,2,((height+4)/2)+floor_height))
              )
    door_frame = door_frame.faces("Z").edges("Y").fillet(((length+8)/2)-.5)
    w = wall.add(door_frame).cut(cutout)
    return w


def make_kitchen():
    bp = Building(length=175, width=175, height=350, stories=2)

    bp.room['build_walls']= [False,True,True,True]
    bp.room['window_walls'] = [False, True, True, False]
    bp.room['door_walls'] = [False, False, False, True]
    bp.room['make_custom_door'] = make_arch_door
    bp.room['wall_width'] = 4
    bp.room['floor_height'] = 3

    bp.door['length'] = 60
    bp.door['height'] = 100

    bp.window['height'] = 55
    bp.window['length'] = 78

    bp.make()
    bp.floors[0].window_count=4
    bp.floors[0].window['height'] = 45
    bp.floors[0].window['length'] = 28
    bp.floors[0].make_custom_windows = casement_windows
    if render_floor:
        bp.floors[0].floor_tile = tile.octagon_with_dots(10, 2.4, 3.2, 1)
        bp.floors[0].floor_tile_padding = 1
    bp.floors[0].make()

    st_front_wall = bp.floors[0].walls[1]
    stone_wall = add_stones(st_front_wall, 175, bp.floors[1].height, bp.floors[1].wall_width)
    bp.floors[0].walls[1] = stone_wall

    st_side_wall = bp.floors[0].walls[2]
    stone_side_wall = add_stones(st_side_wall, 175, bp.floors[0].height, bp.floors[0].wall_width, rotate = -90, seed="test5")
    bp.floors[0].walls[2] = stone_side_wall

    bp.floors[1].make_custom_windows = lattice_windows
    if render_floor:
        bp.floors[1].floor_tile = tile.basketweave(length = 10, width = 5, height = 1, padding = 1)
        bp.floors[1].floor_tile_padding = 1
    bp.floors[1].make()

    front_wall = bp.floors[1].walls[1]
    paneled_wall = add_tudor_frame(front_wall, 175, bp.floors[1].height, bp.floors[1].wall_width, frame_width=5, frame_height=1.5, columns=4, w_length=78, w_height=55)
    bp.floors[1].walls[1] = paneled_wall

    side_wall = bp.floors[1].walls[2]
    paneled_wall2 = add_tudor_frame(side_wall, 175, bp.floors[1].height, bp.floors[1].wall_width, frame_width=5, frame_height=1.5, columns=4, w_length=78, w_height=55, rotate=-90)
    bp.floors[1].walls[2] = paneled_wall2#.rotate((0,0,1),(0,0,0),-90)

    left = bp.build()
    left_roof = make_roof()#x_offset=5)#.translate((5,-5,312.5))
    over_roof = make_dormer_roof(left_roof, 125)
    over_roof2 = over_roof.translate((5,-5,312.5))

    #second_floor = bp.floors[1].build()
    #second_scene = cq.Workplane("XY").add(second_floor)
    #return second_scene

    combine = cq.Workplane("XY").add(left).add(over_roof2)
    return over_roof
    return combine

def make_back_kitchen():
    bp = Building(length=175, width=175, height=350, stories=2)

    bp.room['build_walls']= [True,False,True,True]
    bp.room['window_walls'] = [True, False, True, False]
    bp.room['door_walls'] = [False, False, False, True]
    bp.room['make_custom_door'] = make_arch_door_fancy
    bp.door['length'] = 60
    bp.door['height'] = 100
    bp.room['wall_width'] = 4
    bp.room['floor_height'] = 3


    bp.window['height'] = 55
    bp.window['length'] = 78

    bp.make()
    bp.floors[0].window_count=4
    bp.floors[0].window['height'] = 45
    bp.floors[0].window['length'] = 28
    bp.floors[0].make_custom_windows = casement_windows
    if render_floor:
        bp.floors[0].floor_tile = tile.octagon_with_dots(10, 2.4, 3.2, 1)
        bp.floors[0].floor_tile_padding = 1
    bp.floors[0].make()

    st_front_wall = bp.floors[0].walls[0]
    stone_wall = add_stones(st_front_wall, 175, bp.floors[1].height, bp.floors[1].wall_width, rotate = -180)
    bp.floors[0].walls[0] = stone_wall

    st_side_wall = bp.floors[0].walls[2]
    stone_side_wall = add_stones(st_side_wall, 175, bp.floors[0].height, bp.floors[0].wall_width, rotate = -90, seed="test5")
    bp.floors[0].walls[2] = stone_side_wall

    bp.floors[1].make_custom_windows = lattice_windows
    if render_floor:
        bp.floors[1].floor_tile = tile.basketweave(length = 10, width = 5, height = 1, padding = 1)
        bp.floors[1].floor_tile_padding = 1
    bp.floors[1].make()

    front_wall = bp.floors[1].walls[0]
    paneled_wall = add_tudor_frame(front_wall, 175, bp.floors[1].height, bp.floors[1].wall_width, frame_width=5, frame_height=1.5, columns=4, w_length=78, w_height=55, rotate=-180)
    bp.floors[1].walls[0] = paneled_wall

    side_wall = bp.floors[1].walls[2]
    paneled_wall2 = add_tudor_frame(side_wall, 175, bp.floors[1].height, bp.floors[1].wall_width, frame_width=5, frame_height=1.5, columns=4, w_length=78, w_height=55, rotate=-90)
    bp.floors[1].walls[2] = paneled_wall2#.rotate((0,0,1),(0,0,0),-90)

    left = bp.build()
    left_roof = make_roof().rotate((0,0,1),(0,0,0),180).translate((5,5,312.5))

    combine = cq.Workplane("XY").add(left).add(left_roof)

    second_floor = bp.floors[0].build()
    second_scene = cq.Workplane("XY").add(second_floor)
    return second_scene

    return combine

def make_center():
    bp = Building(length=125, width=175, height=350, stories=2)
    bp.room['build_walls']= [False,True,True,True]
    bp.room['window_walls'] = [False, False, False, False]
    bp.room['door_walls'] = [False, False, True, True]
    bp.room['make_custom_door'] = make_arch_door
    bp.room['wall_width'] = 4
    bp.room['floor_height'] = 4

    bp.door['length'] = 60
    bp.door['height'] = 100

    bp.make()
    bp.floors[0].door_walls = [False, True, True, True]
    bp.floors[0].make()

    front_wall = bp.floors[1].walls[1]
    paneled_wall = add_tudor_frame(front_wall, 125, bp.floors[1].height, bp.floors[1].wall_width, frame_width=5, frame_height=1.5, columns=4)
    bp.floors[1].walls[1] = paneled_wall

    center = bp.build()

    center_roof = make_roof(125)
    over_roof = make_dormer_roof(center_roof, 125)
    over_roof = over_roof.translate((0,-5,312.5))

    combine = cq.Workplane("XY").add(center).add(over_roof)
    #return over_roof
    return combine

def make_center_back():
    bp = Building(length=125, width=175, height=350, stories=2)
    bp.room['build_walls']= [True,False,True,True]
    bp.room['window_walls'] = [False, False, False, False]
    bp.room['door_walls'] = [False, False, True, True]
    bp.room['make_custom_door'] = make_arch_door
    bp.room['wall_width'] = 4
    bp.room['floor_height'] = 4

    bp.door['length'] = 60
    bp.door['height'] = 100

    bp.make()
    bp.floors[0].door_walls = [True, False, True, False]
    bp.floors[0].make()

    bp.floors[1].window_walls = [True,False,False,False]
    bp.floors[1].window_count=3
    bp.floors[1].window['height'] = 55
    bp.floors[1].window['length'] = 25
    bp.floors[1].make_custom_windows = lattice_windows
    bp.floors[1].make()

    front_wall = bp.floors[1].walls[0]
    paneled_wall = add_tudor_frame(front_wall, 125, bp.floors[1].height, bp.floors[1].wall_width, frame_width=5, frame_height=1.5, columns=4, rotate = 180, w_length=78, w_height=55)
    bp.floors[1].walls[0] = paneled_wall

    bp.floors[1].floor.add_operation(test_operation)
    bp.floors[1].floor.make()
    center = bp.build()

    center_roof = make_roof(125, x_offset=-3)
    center_roof = center_roof.rotate((0,0,1),(0,0,0),180).translate((0,5,312.5))

    roof_entrance = cq.Workplane("XY").box(40,30, 20).translate((35,-60,270))
    center_roof = center_roof.cut(roof_entrance)

    stair_lower = stairs(
    length = 148,
    width = 32,
    height = 175,
    run = 8,
    stair_length_offset = 5.35,
    stair_height = 3,
    stair_height_offset = -.8,
    rail_width = 3,
    rail_height = 14,
    step_overlap = None
    )
    stair_lower = stair_lower.rotate((0,0,1),(0,0,0),-90).translate((-15-28,-10,0))

    ladder_bp = Ladder(length=30, height=175, width=8)
    ladder_bp.rung_padding = 12
    ladder_bp.rung_height = 3
    ladder_bp.rung_width = 3
    ladder_bp.make()
    ladder = ladder_bp.build().rotate((0,0,1),(0,0,0),90).translate((55,-60,175))

    combine = cq.Workplane("XY").add(center)
    combine = combine.add(ladder)
    combine = combine.add(stair_lower)
    combine = combine.add(center_roof)

    second_floor = bp.floors[1].build()
    second_ladder = ladder.translate((0,0,-175))
    second_scene = cq.Workplane("XY").add(second_floor).add(second_ladder)
    #return second_scene

    first_floor = bp.floors[0].build()
    first_stair = stair_lower.translate((0,0,0))
    first_scene = cq.Workplane("XY").add(first_floor).add(first_stair)
    return first_scene

    return combine


def make_living():
    bp = Building(length=175, width=175, height=350, stories=2)

    bp.room['build_walls']= [False,True,True,True]
    bp.room['window_walls'] = [False, True, False, True]
    bp.room['door_walls'] = [False, False, True, False]
    bp.room['make_custom_door'] = make_arch_door
    bp.room['wall_width'] = 4
    bp.room['floor_height'] = 4

    bp.door['length'] = 60
    bp.door['height'] = 100

    bp.window['height'] = 55
    bp.window['length'] = 78

    bp.make()

    bp.floors[0].make_custom_windows = casement_windows_2
    bp.floors[0].make()

    st_front_wall = bp.floors[0].walls[1]
    stone_wall = add_stones(st_front_wall, 175, bp.floors[1].height, bp.floors[1].wall_width, seed="test7")
    bp.floors[0].walls[1] = stone_wall

    st_side_wall = bp.floors[0].walls[3]
    stone_side_wall = add_stones(st_side_wall, 175, bp.floors[0].height, bp.floors[0].wall_width, rotate = 90, seed="test9")
    bp.floors[0].walls[3] = stone_side_wall

    bp.floors[1].make_custom_windows = lattice_windows
    bp.floors[1].make()

    front_wall = bp.floors[1].walls[1]
    paneled_wall = add_tudor_frame(front_wall, 175, bp.floors[1].height, bp.floors[1].wall_width, frame_width=5, frame_height=1.5, columns=4, w_length=78, w_height=55)
    bp.floors[1].walls[1] = paneled_wall

    side_wall = bp.floors[1].walls[3]
    paneled_wall2 = add_tudor_frame(side_wall, 175, bp.floors[1].height, bp.floors[1].wall_width, frame_width=5, frame_height=1.5, columns=4, w_length=78, w_height=55, rotate=90)
    bp.floors[1].walls[3] = paneled_wall2#.rotate((0,0,1),(0,0,0),-90)


    right = bp.build()
    right_roof = make_roof()#.translate((-5,-5,312.5))
    over_roof = make_dormer_roof(right_roof, 125)
    over_roof2 = over_roof.translate((-5,-5,312.5))

    combine = cq.Workplane("XY").add(right).add(over_roof2)
    #return over_roof
    return combine

def make_back_living():
    bp = Building(length=175, width=175, height=350, stories=2)

    bp.room['build_walls']= [True,False,True,True]
    bp.room['window_walls'] = [True, False, False, True]
    bp.room['door_walls'] = [False, False, True, False]
    bp.room['make_custom_door'] = make_arch_door_fancy
    bp.room['wall_width'] = 4
    bp.room['floor_height'] = 4

    bp.door['length'] = 60
    bp.door['height'] = 100

    bp.window['height'] = 55
    bp.window['length'] = 78

    bp.make()

    bp.floors[0].make_custom_windows = casement_windows_2
    bp.floors[0].door_walls = [False, False, False, False]
    bp.floors[0].make()

    st_front_wall = bp.floors[0].walls[0]
    stone_wall = add_stones(st_front_wall, 175, bp.floors[1].height, bp.floors[1].wall_width, seed="test7", rotate=180)
    bp.floors[0].walls[0] = stone_wall

    st_side_wall = bp.floors[0].walls[3]
    stone_side_wall = add_stones(st_side_wall, 175, bp.floors[0].height, bp.floors[0].wall_width, rotate = 90, seed="test9")
    bp.floors[0].walls[3] = stone_side_wall

    bp.floors[1].make_custom_windows = lattice_windows
    bp.floors[1].make()

    front_wall = bp.floors[1].walls[0]
    paneled_wall = add_tudor_frame(front_wall, 175, bp.floors[1].height, bp.floors[1].wall_width, frame_width=5, frame_height=1.5, columns=4, w_length=78, w_height=55, rotate=180)
    bp.floors[1].walls[0] = paneled_wall

    side_wall = bp.floors[1].walls[3]
    paneled_wall2 = add_tudor_frame(side_wall, 175, bp.floors[1].height, bp.floors[1].wall_width, frame_width=5, frame_height=1.5, columns=4, w_length=78, w_height=55, rotate=90)
    bp.floors[1].walls[3] = paneled_wall2#.rotate((0,0,1),(0,0,0),-90)

    right = bp.build()
    right_roof = make_roof().rotate((0,0,1),(0,0,0),180)
    right_roof = right_roof.translate((-5, 5,312.5))

    combine = cq.Workplane("XY").add(right).add(right_roof)
    return combine


left = make_kitchen().translate((87.5 + 62.5,0,0))
#left_back = make_back_kitchen().translate((87.5 + 62.5,175,0))
#center = make_center()
#center_back = make_center_back().translate((0,175,0))
#right = make_living().translate((-87.5 - 62.5,0,0))
#right_back = make_back_living().translate((-87.5 - 62.5,175,0))

scene = (cq.Workplane("XY")
         .add(left)
         #.add(left_back)
         #.add(center)
         #.add(center_back)
         #.add(right)
         #.add(right_back)
         )

#baskt = tile.basketweave(length = 8, width = 4, height = 1, padding = 1)
#show_object(baskt)
#bounds = baskt.val().BoundingBox()
#t_width = bounds.ylen
#t_length = bounds.xlen
#t_height = bounds.zlen
#log(t_width)
#log(t_length)
#log(t_height)
#columns = math.floor(175/(t_width+ 1))
#rows = math.floor(175/(t_length +1))
#log(columns)
#log(rows)


if cq_editor_show:
    show_object(scene)

if export_to_file:
    cq.exporters.export(scene,'out/dollhouse_10_kitchenRoof.stl')
